### PR TITLE
Softmax ocl refactoring

### PR DESCRIPTION
**Pull request by Vsevolod1983**
TICKET:
https://github.com/ROCm/MIOpen/issues/2643

SoftmaxForward and SoftmaxBackward are refactored in this PR
Also, network config is changed as well, according to what we discussed with the team

### CODE REVIEW

**Review by DrizztDoUrden — APPROVED** [review]
LGTM

File: src/include/miopen/ctc.hpp
Comment (CAHEK7) [inline src/include/miopen/ctc.hpp:59]: Is it related to the softmax?

File: src/ocl/ctcocl.cpp
Comment (CAHEK7) [inline src/ocl/ctcocl.cpp:47]: Is it related to the softmax?

File: src/include/miopen/softmax/invoke_params.hpp
Comment (CAHEK7) [inline src/include/miopen/softmax/invoke_params.hpp]: clang-format

File: src/include/miopen/softmax/problem_description.hpp
Comment (CAHEK7) [inline src/include/miopen/softmax/problem_description.hpp]: Could you add that kind of check?
https://github.com/ROCm/MIOpen/blob/5e82b808021308040d6651e0f759d9a50702c859/src/ocl/softmaxocl.cpp#L78

And also verify `alpha_` and `beta_` pointers for `nullptr`?

**Review by CAHEK7 — CHANGES_REQUESTED** [review]
Some minor changes are needed.

File: src/include/miopen/ctc.hpp
Comment (Vsevolod1983) [reply src/include/miopen/ctc.hpp:59]: Yes it does. CTC uses softmax and passes handle for softmax calculation. Since this handle is no longer a const for softmax we need to remove const modifier here as well. If we have other options lets discuss.

File: src/include/miopen/softmax/problem_description.hpp
Comment (Vsevolod1983) [reply src/include/miopen/softmax/problem_description.hpp]: These checks  are present in  miopenStatus_t SoftmaxForward(...) function
But I will add additional alpha and beta checks.

Comment (atamazov) [reply src/include/miopen/softmax/problem_description.hpp]: @Vsevolod1983 It seems better to check and block construction of invalid ProblemDescriptions rather than implement the same checks here and there in the code. So please,

> These checks are present in miopenStatus_t SoftmaxForward(...) function

Move these into ctor of PD. Test manually if it works properly.

File: src/include/miopen/ctc.hpp
Comment (atamazov) [reply src/include/miopen/ctc.hpp:59]: @Vsevolod1983 is it so that we need to make many other changes in the code in order to keep the handle `const`?

Comment (atamazov) [inline src/include/miopen/ctc.hpp:59]: @Vsevolod1983 is it so that we need to make many other changes in the code in order to keep the handle `const`? If yes, and we are not going to do this in this PR, then let's open a ticket and eventually do this in a followup PR.

/cc @CAHEK7

**Comment by CAHEK7** [comment]
I guess it can be converted from draft in order to trigger CI.

File: src/include/miopen/softmax/problem_description.hpp
Comment (Vsevolod1983) [reply src/include/miopen/softmax/problem_description.hpp]: @atamazov 
We cannot do it in all cases, because several parameters are not part of ProblemDescription. For example x and y - are data params, they are not included in ProblemDescription ctor, but we need to check them as well (and throws an exception if they are invalid).
Also, we have a case when we throw an exception only after we calculate a lot of intermediate params. So, its not a part of PD ctor as well.

File: src/include/miopen/ctc.hpp
Comment (Vsevolod1983) [reply src/include/miopen/ctc.hpp:59]: CTCLoss calls SoftmaxForward which subsequently calls SolverContainers::ExecutePrimitive which accept non const Handle&.
It leads to CTCLoss should accept non const Handle&. Otherwise we need to use const-cast hack.

Comment (atamazov) [reply src/include/miopen/ctc.hpp:59]: @Vsevolod1983 I see, thanks.

@DrizztDoUrden What about https://github.com/ROCm/MIOpen/pull/1276#discussion_r1455119631?

File: src/include/miopen/softmax/problem_description.hpp
Comment (atamazov) [reply src/include/miopen/softmax/problem_description.hpp]: @Vsevolod1983 
> Also, we have a case when we throw an exception only after we calculate a lot of intermediate params. So, its not a part of PD ctor as well.

Does this case indicate invalid problem?

**Comment by junliume** [comment]
> I guess it can be converted from draft in order to trigger CI.

CI started, I also requested the user names to be added to the repo access for more convenient development process.

Update @Vsevolod1983 there are some HIP tidy issues:
[hip_tidy.log](https://github.com/ROCm/MIOpen/files/13970045/hip_tidy.log)

**Comment by Vsevolod1983** [comment]
I also need to remake network config as we discussed with the team. Working on it. So please do not merge for now

File: src/include/miopen/softmax/problem_description.hpp
Comment (Vsevolod1983) [reply src/include/miopen/softmax/problem_description.hpp]: @atamazov 
Hi, I moved some checks to ctor of PD class. One check stayed on solver side though. Please have a look.

Also I changed network config, will appreciate if you have a look as well.

**Comment by Vsevolod1983** [comment]
Its ready to review now

**Comment by Vsevolod1983** [comment]
> > I guess it can be converted from draft in order to trigger CI.
> 
> CI started, I also requested the user names to be added to the repo access for more convenient development process.
> 
> Update @Vsevolod1983 there are some HIP tidy issues: [hip_tidy.log](https://github.com/ROCm/MIOpen/files/13970045/hip_tidy.log)

investigating, thank you
UPD: Fixed

File: src/include/miopen/softmax/problem_description.hpp
Comment (atamazov) [inline src/include/miopen/softmax/problem_description.hpp]: [Q] Maybe it is better to initialize it with an empty tensor descriptor? You can create it as a static member of the class and name "unusedDesc".

File: src/include/miopen/softmax/invoke_params.hpp
Comment (atamazov) [inline src/include/miopen/softmax/invoke_params.hpp]:
```suggestion
    // xdxDesc is used for both forward and backward
```
typo here and below

File: src/include/miopen/softmax/invoke_params.hpp
Comment (atamazov) [inline src/include/miopen/softmax/invoke_params.hpp]: [Q] Maybe it is cleaner to follow the same pattern as used for yDesc? (xDesc, forward_x, backward_x)?

[Q] Or perhaps it is worth creating separate Problems for Forward and Backward? Would this result in improved code clarity?

This is not very important right now, but anyway...

File: src/solver/softmax/softmax.cpp
Comment (atamazov) [inline src/solver/softmax/softmax.cpp:189]: [Quality] Magic numbers. It this is a legacy stuff, you can simply mark them with `/// \todo Magic numbers` (but of course it is better to fully resolve this review comment).

File: src/solver/softmax/softmax.cpp
Comment (atamazov) [inline src/solver/softmax/softmax.cpp:195]: Ditto

File: src/solver/softmax/softmax.cpp
Comment (atamazov) [inline src/solver/softmax/softmax.cpp:147]: [Q] Maybe it is better to make GetAlpha/Beta() return floating-point numbers?

File: src/solver/softmax/softmax.cpp
Comment (atamazov) [inline src/solver/softmax/softmax.cpp:88]: [Quality] Magic numbers. It this is a legacy stuff, you can simply mark them with `/// \todo Magic numbers` (but of course it is better to fully resolve this review comment).

File: src/solver/softmax/softmax.cpp
Comment (atamazov) [inline src/solver/softmax/softmax.cpp:102]: Ditto

**Review by atamazov — COMMENTED** [review]
Some questions + comments.

This can be merged provided that performance testing results (baseline vs this PR) are good.

**Comment by atamazov** [comment]
@Vsevolod1983 Good work! What about performance testing?

**Comment by CAHEK7** [comment]
> Some questions + comments.
> 
> This can be merged provided that performance testing results (baseline vs this PR) are good.

Just a sidenote:
since GPU part has not been changed - the only affected part is a host part.

**Review by CAHEK7 — APPROVED** [review]

**Comment by CAHEK7** [comment]
@atamazov could you help with perf measurement methodology?
It's not quite obvious what should be measured for this particular set of changes.

File: src/include/miopen/softmax/problem_description.hpp
Comment (CAHEK7) [reply src/include/miopen/softmax/problem_description.hpp]: I would ask more generic question - what is the lifetime of all the references? Can we guarantee that it won't be constructed from temporaries?

I'm not quite sure that lifetime extension will work in that case:
```C++
ProblemDescriptor pd(TensorDescriptor());
auto& td_red = pd.getTensortDescriptor();
```

Comment (CAHEK7) [reply src/include/miopen/softmax/problem_description.hpp]: It is potentially unsafe.
https://godbolt.org/z/vnfM9EYv7

**Review by CAHEK7 — CHANGES_REQUESTED** [review]

File: src/softmax.cpp
Comment (CAHEK7) [inline src/softmax.cpp]:
```suggestion
```

File: src/include/miopen/softmax/problem_description.hpp
Comment (Vsevolod1983) [reply src/include/miopen/softmax/problem_description.hpp]: I will change this on using just Tensors but not const references to tensors.

File: src/solver/softmax/softmax.cpp
Comment (Vsevolod1983) [reply src/solver/softmax/softmax.cpp:88]: legacy, will put a comment

File: src/include/miopen/softmax/invoke_params.hpp
Comment (Vsevolod1983) [reply src/include/miopen/softmax/invoke_params.hpp]: Q1: Could you please clarify, how exactly you suggest to make a change?
Q2: I thought about is it better to separate classes or not and concluded its better to have one class, because functionality is really similar. Also, I checked implementation for other operations and saw that this approach is quite commoin. 
I introduced two constructors for more handy usage.

**Comment by Vsevolod1983** [comment]
> @Vsevolod1983 Good work! What about performance testing?

I reworked PD and  InvokeParams and now using TensorDescriptor instead of const ref to it
Also, put changes regarding the rest of the comments(vast majority of them)

**Comment by Vsevolod1983** [comment]
@atamazov 
image
Is it ok ?

**Comment by atamazov** [comment]
@Vsevolod1983 Pictures? 🤣 
> Is it ok ?

Who knows? In order to evaluate how performance has changed, the one needs to compare the baseline vs patch. What are the baseline (currently, this is `develop` at commit 959a7aa0803e64751f36b036795e797fb03563c8) numbers? Let's have a call tomorrow (or on Monday).

File: src/solver/softmax/softmax.cpp
Comment (atamazov) [reply src/solver/softmax/softmax.cpp:102]: [resolved]

File: src/solver/softmax/softmax.cpp
Comment (atamazov) [reply src/solver/softmax/softmax.cpp:88]: [resolved]

File: src/include/miopen/softmax/invoke_params.hpp
Comment (atamazov) [reply src/include/miopen/softmax/invoke_params.hpp]: Ok [resolved]

File: src/include/miopen/softmax/problem_description.hpp
Comment (atamazov) [reply src/include/miopen/softmax/problem_description.hpp]: @Vsevolod1983 Oops, yeah, we need copies of TDs. Let's follow the pattern used for convolutions.

**Review by CAHEK7 — APPROVED** [review]
Seems fine. Lets take a look at perf measurements.

**Comment by Vsevolod1983** [comment]
Compared PR time on softmax with develop's branch time. So if talk about so called Wall Time we have slight time execution increase on PR which is quite expected.
Wall Time diff is apprx 0.5% for SoftMaxbackward and 2.4% for SoftmaxForward. However, it's worth to mention that it slightly differs from run to run, which is probably also expected.

File: src/include/miopen/softmax/invoke_params.hpp
Comment (junliume) [inline src/include/miopen/softmax/invoke_params.hpp:136]: @Vsevolod1983 This line and the Line 141 below:
```
[2024-01-30T19:37:27.888Z] /home/jenkins/workspace/MLLIBS_MIOpen_PR-2671/src/include/miopen/softmax/invoke_params.hpp:136:21: error: implicit conversion 'float' -> bool [readability-implicit-bool-conversion,-warnings-as-errors]

[2024-01-30T19:37:27.888Z]             alpha = *(static_cast<const float*>(alpha_));

[2024-01-30T19:37:27.888Z]                     ^

[2024-01-30T19:37:27.888Z]                     (                                    != 0.0f)
```

**Comment by atamazov** [comment]
## 🌀 Performance testing results

Preconditions:
- Navi21 (gfx1030/36)
- ROCm 6.0 release docker, Ubuntu 20.04.5 LTS
- BUILD_DEV=Off, Release build
- develop ad21ccf7e vs. branch 76057d8e8
- 35 known FP32 configs (borrowed from my convolution tests), tensor sizes vary from 10k to ~2.9M elements
- Binary cache pre-filled with kernels, `-i 1000` used to neutralize the effect of loading kernels from the cache. This is necessary because the softmax driver neither implements the "Aux Wall Time" measurements nor the "warm-up" function.

Results:
- 🟢 GPU Time - jitter around 0.2% (expected)
- 🟢 Wall Time (includes GPU time) - average drop around 0.8%, max drop 6%
  - [Informative] Average drop for Forward is 0.6%, for Backward is 0.9%. This doesn't match the results listed at https://github.com/ROCm/MIOpen/pull/2671#issuecomment-1915292066. I'll share the methodology and scripts with Vsevolod (and everyone who wants to get this too).
- 🟡 Host Overhead Time (computed as `wall_time - gpu_time`) - average drop 4.5%, max drop 9%

Good to go!

File: src/softmax/problem_description.cpp
Comment (atamazov) [inline src/softmax/problem_description.cpp:71]: [Notice] 💡 You can use EOL comments to enforce structured formatting, like this:
```C++
network_config += "n_y" + std::to_string(n_y) //
        + "c_y" + std::to_string(c_y) //
        + "h_y" + std::to_string(h_y) //
        + "w_y" + std::to_string(w_y) //
...
```
clang-format will do the rest of the job.

File: src/softmax/problem_description.cpp
Comment (atamazov) [inline src/softmax/problem_description.cpp:69]: [Notice] No empty lines please. But you can use EOL comment in empty line to emphasize that it is left empty intentionally.
```C++
       network_config += "n_y" + std::to_string(n_y) + "c_y" + std::to_string(c_y) + "h_y" +
                          std::to_string(h_y) + "w_y" + std::to_string(w_y) +
                          //
                          "n_dy" + std::to_string(n_dy) + "c_dy" + std::to_string(c_dy) + "h_dy" +
                          std::to_string(h_dy) + "w_dy" + std::to_string(w_dy) +
```
but I do not know how clang-format will format this.

File: src/softmax/problem_description.cpp
Comment (atamazov) [inline src/softmax/problem_description.cpp:48]: Rewrite this stuff using `ostringstream`, see https://github.com/ROCm/MIOpen/blob/c7fa6f7fbec5f2a76128e2c9ae80f66c802722a6/src/conv/problem_description.cpp#L106

**Review by atamazov — CHANGES_REQUESTED** [review]
Use string streams instead of merging strings. Some formatting tips.

File: src/softmax/problem_description.cpp
Comment (Vsevolod1983) [reply src/softmax/problem_description.cpp:48]: @atamazov 
Merging strings in this style were in initial code which is in develop currently. Scope of this ticket was to introduce solver infrastructure for softmax. Could we improve this in additional ticket / PR ? 

Clang-format was run on this PR.

Comment (atamazov) [reply src/softmax/problem_description.cpp:48]: @Vsevolod1983 The "problem description" class is introduced in this PR, so I would prefer fixing right here. However, I won't insist if you would like to postpone this for a while.

**_This PR can be merged as is provided that the next your PR is the one that resolves this review comment._**

**Review by atamazov — APPROVED** [review]
🟢 LGTM

File: src/softmax/problem_description.cpp
Comment (atamazov) [reply src/softmax/problem_description.cpp:48]: Agreed to make the fix in the next PR.

Comment (Vsevolod1983) [reply src/softmax/problem_description.cpp:48]: Discussed with @atamazov . PR can be merged.

**Review by junliume — APPROVED** [review]

File: src/solver/softmax/softmax.cpp
Comment (atamazov) [inline src/solver/softmax/softmax.cpp:286]: [Performance] @Vsevolod1983 ⚠️ **This review is written with an assumption that we are passing alpha and beta as kernel arguments, _AND intend to do so in the future._** 

If the above is true, then it does not make sense to create different Invoker instances for different alpha and beta, because kernel binaries remain the same for different alpha and beta. We are simply wasting time.

- Note: I am wondering if this is one of the reasons of 4.5% Host Overhead Time regression mentioned at https://github.com/ROCm/MIOpen/pull/2671#issuecomment-1918034107.

Yes, we _can_ include alpha and beta in PD (ProblemDescription), but they should be of no use there. Specifically:
- thou shalt not serialize them into cache/db keys, including keys for the Invoker cache. 
- thou shalt not use these values in IsApplicable and other solver functions.
- thou shalt not capture them in the InfokerFactory.

The only use of alpha and beta is passing via InvokeParams.

Right here: take alpha and beta values from params:
```C++
                kernel(params.x,
                       ...
                       params.alpha,
                       params.beta);
```
plus some other changes (details omitted).

/cc @CAHEK7 @DrizztDoUrden

**Review by atamazov — COMMENTED** [review]
Let's use the same Invoker for different alpha/beta.

**Pull request merged**
